### ISSUE TITLE
DOS version of DOSBox-X does not run in DOSBox-X

### ISSUE THREAD

**Issue by Wengier** [joncampbell123/dosbox-x#1830]
As a DOS emulator, DOSBox-X should be able to run its DOS version as well, but it turns out not to be the case. Tried bigger memory size (e.g. 64MB), UNIVBE to add VESA modes, and running in a DOS guest system (instead of the integrated DOS), neither will work - it will either crash or show a black screen with mouse depending on the HXGUIHLP.INI setting. Tried it in DOSBox ECE and dosbox-staging, they will not work either, but (unlike DOSBox-X) they are primarily for running DOS games. I believe the problem is likely related to the VESA code.

**Comment by grapeli** [comment]
Linux version works. No revelation.
It also works under dosbox-SVN and dosbox-staging.

Older version, because I'm lazy and unwilling to build a compiler newer than gcc-4.7.4 (its compatibility with -std=c++11 is not enough). No sound support, no mouse or anything else. So unnecessary for testing purposes.

memsize=63 should usually do.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5191007/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
I tried your package, and it works indeed. Thanks for this. But it is not the official DOS package I was talking about, and it uses a completely different approach by booting into Linux via LOADLIN, yet there is no way to boot back to DOS unless you completely restart DOSBox-X. It is mostly useful for demonstration purpose I guess.

**Comment by grapeli** [comment]
Yes, this is a purely experimental version, probably with many bugs.

As for the dosbox-x version for HX-DOS. I was running it under Qemu and dosemu2 (it works). Under Qemu with freedos requires at least 87-90MB of memory (quite a lot) to run.

Under dosbox-x it starts to boot after changing `core=normal` and modifying the HXGUIHLP.INI file (change to 640x480x8) and seems to stop. Maybe it requires a higher resolution, although under qemu and dosemu2 this is fine.

**Comment by Wengier** [comment]
@joncampbell123 I see the 0.83.6 release note currently says "hx-dos builds are not available due to symbols that do not exist when compiled for that target.". Which symbol(s) do not exist for HX-DOS? Perhaps I can fix it quickly.

**Comment by joncampbell123** [comment]
tinyfd_MessageBox. I'll pull up the Windows XP VM and compile again to get the error message.

I am compiling MinGW builds now, try not to change code during this time please.

**Comment by joncampbell123** [comment]
Here you go.
[Errors.txt](https://github.com/joncampbell123/dosbox-x/files/5309948/Errors.txt)

**Comment by Wengier** [comment]
@joncampbell123 I think I know why. Not difficult to fix, but I will wait for you to finish the current compilation of course.

**Comment by Wengier** [comment]
@joncampbell123 Please let me know when it is ready to push the code. There is only one file to be slightly changed.

**Comment by joncampbell123** [comment]
@Wengier Fell asleep early, go ahead and push for HX-DOS. I'll upload MinGW64 and Mac OS X builds that finished last night.

**Comment by Wengier** [comment]
@grapeli The latest code already supports the x86 dynamic core (both 32-bit and 64-bit, with ```core=dynamic``` or ```core=dynamic_x86```), so it will hopefully be noticeable faster than before. Also, there is now official support for Linux Flatpak, so Linux support is also improved. But I wonder if there is a way you can boot back to DOS once you start the DOSBox-X Linux version from DOS? Since the current HX-DOS package does not run inside DOSBox-X, I think the Linux approach can indeed be a pretty good alternative approach if major issues are solved. In any case, thanks for demonstrating that such a version does indeed work!

**Comment by grapeli** [comment]
@Wengier 
Before that, I was interested in the comparison with dosbox-SVN and staging (universal build). They cannot handle the keyboard on kernels newer than 2.4 (8042 auxiliary port). A serious limitation.
In general, under dosbox, it is only possible to boot the linux kernel under 386 architecture. On Linux, support for it was removed in 2013 (newer than 3.7 won't work).
https://kernelnewbies.org/Linux_3.8#Removal_of_support_for_386_processors

I will try to build the latest dosbox-x.

**Comment by grapeli** [comment]
The latest dosbox-x with fluidsynth support.
I have configured the kernel with OSS support, including SDL, but I can't get them to work.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5453990/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Sure, you can check and compare these branches (you are apparently an experienced user). We are also trying to helping each other and several collective efforts are made as well. Recent examples include  https://github.com/joncampbell123/dosbox-x/issues/1961 and https://github.com/joncampbell123/dosbox-x/issues/1964.

Back to topic - I have checked out the latest DOSBox-X Linux build you made. Thanks for making it, and it certainly runs successfully inside DOSBox-X itself. Since it appears to run better with more DOS file handles, I have now increased the default maximum DOS handles in the latest DOSBox-X code, although it is always customizable in the config file (in [config] section which tries to emulate DOS's CONFIG.SYS file).

The other issue was that there is no way to boot back to DOS once LOADLIN has started - even the menu items "Reset virtual machine" or "Reboot guest system" apparently did not work previously. So I have decided to fix the resetting function as well, and with the latest code I can now click the reboot menu option to boot back to DOS, at least with core=normal or core=auto. I do not know if there is a native way to exit from LOADLIN, but getting DOSBox-X's reboot function to work for LOADLIN (as already done) is certainly very helpful.

I think the biggest issue for now seems that there is no mouse support, so I cannot select the menu items from DOSBox-X running inside LOADLIN. Additionally, moving the mouse may sometimes cause DOSBox-X inside LOADLIN to freeze, although I can now select DOSBox-X's reboot function to restart LOADLIN. I wonder if there is a way to get the mouse to work?

I have also tried to play the /root/dos/DOSMID/12.mid file with DOSMID inside LOADLIN+DOSBox-X, and it does show the graphical interface and seem to be playing, but I did not hear sound yet. Perhaps there are some settings (or other adjustments) required to get the sound to work?

In any case, it is definitely getting better and better, and I highly appreciate your effort.

**Comment by grapeli** [comment]
@Wengier 
A [fifo](https://github.com/joncampbell123/dosbox-x/blob/master/vs2015/sdl/src/video/fbcon/SDL_fbevents.c#L55) `/dev/gpmdata` was missing. Now the mouse works in the menu (as far as).
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5457836/dosbox-x.linux.zip)

Fluidsynth works properly under qemu only with kvm acceleration. With tcg, also not enough power. Switch to the second console (alt-f1) and run top.
```
qemu-system-x86_64 -machine pc,accel=kvm -cpu host -m 128 -net none \
-device sb16,iobase=0x220,irq=7,dma=1,dma16=5 \
-kernel /tmp/dosbox-x/vmlinuz \
-initrd  /tmp/dosbox-x/dosbox.cfs \
-append 'ramdisk_size=12288K root=/dev/ram0 ro vga=788 sb=0x220,7,1,5 dmabuf=1 quiet' -monitor stdio
```
Too bad dosbox-x doesn't work with the linux 8-bit framebuffer. Framebuffer is generally slow with a 16-bit color palette much more.

qemu-system-x86_64 -machine pc,accel=kvm -cpu host -m 128 -net none -device sb16,iobase=0x220,irq=7,dma=1,dma16=5 -kernel /tmp/dosbox-x/vmlinuz -initrd  /tmp/dosbox-x/dosbox.cfs -append 'ramdisk_size=12288K root=/dev/ram0 ro vga=788 sb=0x220,7,1,5 dmabuf=1 quiet' -monitor stdio

**Comment by grapeli** [comment]
@Wengier 
Latest build containing this [commit](https://github.com/joncampbell123/dosbox-x/commit/8f95e751ecb06486080546eff358c43fc7849a57).
Compiled with flags, among others `-march=i386 -mtune=generic` .

dosbox-x, running linux, top, ~10% CPU usage
qemu (accel=tcg), running linux, top, 0.2-04% CPU usage

**Comment by Wengier** [comment]
@grapeli Yes, I can now use the mouse to activate the menu with the new build, which is certainly a good progress, although the mouse operations may appear to be somewhat buggy. But I can open for example Configuration Tool and Mapper Editor from menu just fine, so the menu is already starting to become usable.

Since I am primarily using Windows, kvm is not available for QEmu on Windows. I usually run Linux inside a VM, and apparently you cannot expect a good speed for this. When I run the Linux DOSBox-X inside Windows DOSBox-X Fluidsynth did not appear to work yet, but OPL3 did work so that "dosmid /opl 12.mid" will indeed play the music (although speed may be a concern).

For Linux 8-bit framebuffer support, I am not sure if there is an existing config option in DOSBox-X for this purpose. Perhaps @joncampbell123 knows more about this.

As for CPU usage, perhaps you can try DOSBox-X's built-in DOSIDLE program. Activating it may significantly reduce the CPU usage.

Once again, thanks for the improvement. I hope to further improve DOSBox-X's compatibility with LOADLIN as well.

**Comment by joncampbell123** [comment]
@Wengier The way the code is written (and DOSBox SVN has the same limitation) makes it extremely unlikely DOSBox/DOSBox-X would ever support rendering to a 256-color display.

**Comment by grapeli** [comment]
@Wengier 
Fluidsynth is currently not working.
```
fluisynth: error: Device </dev/dsp> could not be opened for writing: Device or resource busy
LOG: MIDI: fluidsynth: Can't create audio driver
```
I am 100% sure that it worked before. Soundfont was also loaded (I saw it in the logs). I'll check it (I don't know what's going on). It works.
`/bin/fluidsynth -m oss -a oss -l -i /etc/sf2/minipiano.sf2 /root/dos/DOSMID/12.mid`

As for the support of 256 colors. Maybe there is some incompatibility with the linux framebuffer. I checked an earlier HX-DOS version of dosbox-x and it worked fine in that color space.

I only treat it as a curiosity. I don't see any use in this.

**Comment by Wengier** [comment]
@grapeli I see. It is certainly helpful to get Fluidsynth working again.

Meanwhile, I have tried to further improve compatibility with LOADLIN - now menu items such as "Show Sound Blaster configuration" and "Show MIDI device configuration" should work in the *outside* DOSBox-X that has LOADLIN running. Previously it did not appear to work at all and will likely cause a LOADLIN kernel panic message.

**Comment by grapeli** [comment]
> I see. It is certainly helpful to get Fluidsynth working again.
> 

I can use ALSA. It can mix sound from multiple sources (dmix) when the sound card has no hardware mixer. The libasound library is linked in dosbox-x anyway. Useless. Although SDL and fluidsynth were built without alsa support, dosbox-x requires libasound for linking. Although it was configured with `--disable-alsa-midi`.

I wanted to simplify the configuration, minimize the size of the binary code.

I will build another toolchain based on uclibc-ng or musl-libc. Maybe it will be a better option. It certainly won't provide as good binary compatibility as this one (Linux 2.4.6).

**Comment by grapeli** [comment]
New build. In one better, in another, weaker.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5469929/dosbox-x.linux.zip)

The menu works correctly after clicking on **Video>Toogle fullscreen**.
The sound works. Fluidsynth too.

Framebuffer (vesafb) is slower. Much better under dosbox-x, [s3fb](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/tree/drivers/video/fbdev/s3fb.c?h=linux-4.4.y) does not work 100% correctly with dosbox on this kernel (I had to drop it).
Poweroff and reboot doesn't work. Busybox (default config) is very truncated. A few redundant libraries, pcre, glib2 (fluidsynth can be built without these dependencies).

Under qemu it runs at a speed a few percent slower than the native binary.
`qemu-system-x86_64 -machine pc,accel=kvm -cpu host -m 128 -net none -device sb16,iobase=0x220,irq=7,dma=1,dma16=5 -kernel vmlinuz -initrd dosbox.cfs -append 'root=/dev/ram0 rw vga=788 ide-core.noprobe=0.0 ide-core.noprobe=0.1 ide-core.noprobe=1.0 ide-core.noprobe=1.1 quiet' -monitor stdio
`

**Comment by Wengier** [comment]
@grapeli Yes, the latest build of DOSBox-X Linux runs fine inside my latest DOSBox-X Windows with the settings memsize=63 and files=255. The menu works too (in full-screen mode as you mentioned). So it is already usable, and I can run the programs in the Linux directories.

For some reason mididevice=fluidsynth does not work here - when I click the menu "Show MIID device configuration" it says "MIDI device: None". But think it works on your system I guess it may be a configuration issue here.

Overall, it is pretty impressive, and thanks for the efforts!

**Comment by grapeli** [comment]
@Wengier 
Memory usage. The key is kernel configuration. The current one is underdeveloped (Is bad). The kernel boots at `memsize=32` and with this initrd, it will be at 16-18. Counting down the memory needed for dosbox.cfs is too much.

The file system is read-only. You can modify the configuration file.
`vi .dosboxrc`

The whole thing is makeshift. To perfect it, you need to have knowledge, appropriate skills and a lot of time (I miss these three elements).

**Comment by Wengier** [comment]
@grapeli I don't think there is actually a need to "perfect" it anyway. You can never expect any DOS version (HX-DOS or LOADLIN) of DOSBox-X to work the same way as if you are running the Windows or Linux version of DOSBox-X. In fact, many features have to be disabled in the HX-DOS build so that it will build successfully. However, since the main target users for the DOS package are DOS users (MS-DOS or FreeDOS users for example), they will definitely know even if the DOS version will run on their DOS systems, it cannot provide all features as in the other OSes.

Considering the above, I think the general expectation of any DOS version of DOSBox-X is basically this - it will launch reliably from a DOS environment (real DOS or inside DOSBox-X in the case of LOADLIN version), and from its shell users can run basic DOS programs, and they can also use the major menu functions (Configuration Tool etc) or change the common config options (frequently-used machine types etc). On the other hand, they cannot expect all menu functions/config options/features to work the same way as in Windows or Linux version of DOSBox-X. If any DOS package can meet this expectation then I think it can be considered a functioning DOS package.

**Comment by Wengier** [comment]
The HX-DOS package has the advantage that it is actually running in DOS in the whole time, and users can exit to DOS any time they want. On the other hand, the LOADLIN version cannot do this, so it is more for running in an emulator like DOSBox-X itself. The HX-DOS package is a truer DOS package in this sense, but the LOADLIN version may be considered an alternative solution.

**Comment by grapeli** [comment]
@Wengier 
New edition. I added s3fb (S3 Trio frame buffer). Initialization slows down booting.
Its advantage is scaling.
There are problems with the menu, it locks up when selected **Video>Fit to aspect ratio**.

Dosbox is launched after startup.. It takes a while. `dosbox.cfs` is [compressed](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/filesystems/cramfs.rst).

**Comment by Wengier** [comment]
@grapeli Thanks for the updates. I have tested it but I think this new version has display issues - when it runs the text on the screen may suddenly disappear from the Linux shell, and sometimes I simply got the following screen instead:

![image](https://user-images.githubusercontent.com/8216923/97937505-e9132c80-1d4c-11eb-84a4-7f8e276705fd.png)

I run it with ```-set memsize=63 -set files=255 -set aspect=true'''. On the other hand, previous editions do not seem to have this problem.

**Comment by grapeli** [comment]
@Wengier 
It works for me. Both in dosbox-x and in qemu (either directly or from freedos).
Copy the previous kernel (vmlinuz), and in linux.par change these two lines to.
```
vga=788
#video=s3fb

```

**Comment by Wengier** [comment]
@grapeli Yes, I can get the latest binary working fine with ```memsize=127```. No other settings are required although it may work somewhat better in full-screen mode.

I think there is one thing that may be somewhat confusing - after I run LINUX.BAT it will show the message "dosbox-x login: " as if it requires the user to type the login information. But as I type any string the the screen text will soon automatically disappear and after a while the Linux DOSBox-X interface will show up. So why not change the message "dosbox-x login: " to something like "Loading DOSBox-X..." so that the user will know that DOSBox-X is actually being automatically loaded instead of thinking that they may need to type some information after the message?

Overall it is pretty impressive I think.

**Comment by Wengier** [comment]
@grapeli The DOSBox-X DOS LOADLIN package is now available from:

* [dosbox-x-dos-0.83.7-loadlin.zip](https://github.com/joncampbell123/dosbox-x/files/5490637/dosbox-x-dos-0.83.7-loadlin.zip)

It is based on your last build, but I renamed LINUX.BAT to DOSBOX-X.BAT and also added a README.TXT file, which says the following:

```
README for DOSBox-X DOS LOADLIN package
=======================================

This is the alternative DOS package for DOSBox-X 0.83.7 using the LOADLIN approach.
Instead of the official HX-DOS approach, you can also run DOSBox-X from a DOS environment using the LOADLIN approach.
This approach is made possible thanks to the work of grapeli.

With this alternative DOS package you can even run DOSBox-X right from DOSBox-X's DOS shell.
Start the outside DOSBox-X with the setting "memsize=127" (and perhaps also "fullscreen=true").
Go to the directory where the files are extracted and type "DOSBOX-X".
Then just wait for DOSBox-X to be automatically loaded within DOSBox-X.
```

I also added the information to the [INSTALL](https://github.com/joncampbell123/dosbox-x/edit/master/INSTALL.md) page offering it as an alternative DOS package so that other users can use it if they want. Thanks a lot for your work, and feel free to let me know if you made any further improvements.

**Comment by grapeli** [comment]
In my opinion, it's too early.

I quickly cleaned up the `/etc/init.sh` file (some nonsense). Now dosbox-x automatically loads the configuration file on startup after login (root).
Check if it works and repack it.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5491241/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Sure, the new version works. The LOADLIN approach gives the possibility of running DOSBox-X inside DOSBox-X, so it is certainly useful and can coexist with the HX-DOS approach. I guess it is already in a usable state because of the recent improvements (both to the package itself and to DOSBox-X's LOADLIN compatibility). I have already repackaged it based on the new version (as [dosbox-x-dos-0.83.7-loadlin.zip](https://github.com/joncampbell123/dosbox-x/files/5491386/dosbox-x-dos-0.83.7-loadlin.zip)), but please feel free to make any further improvements. Thanks for the work!

**Comment by grapeli** [comment]
@Wengier 
This time I was compiling with PGO. Dosbox-x has a smaller size, runs faster, but starts much slower. 
I will try FDO profiling (perf). Maybe it will give a better effect.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5501425/dosbox-x.linux.zip)

> I think there is one thing that may be somewhat confusing - after I run LINUX.BAT it will show the message "dosbox-x login: " as if it requires the user to type the login information.
> 
I have no idea for a good solution to this yet.

**Comment by Wengier** [comment]
@grapeli Thanks for the improvements. Yes it runs faster, but starts slower as you said. I hope to see it is becoming better too.

Meanwhile, a user asked a few questions for the posted LOADLIN DOSBox-X package in another forum (see [here](http://www.bttr-software.de/forum/forum_entry.php?id=17125)) . His questions are the following:

```
How do I mount disk drives in this version? It cannot access my DOS disks.
When I mounted the root I cannot see any /dev/sdx or hdx devices to mount.
What kernel does it use? Would work native with my SB Audigy?
Also it has some menu drawing problem that some block of menu background are drawn to the top of screen and doesn't cleared when menu is closed.
```

He is a real DOS (MS-DOS/FreeDOS) user trying to run this package. Do you happen to know the answers to his questions? Thanks!

**Comment by grapeli** [comment]
@Wengier 
I created this to test run DOSBox-X in DOSBox-X. I also check it for Qemu (a better build could be created strictly for it).

> How do I mount disk drives in this version? It cannot access my DOS disks.
> When I mounted the root I cannot see any /dev/sdx or hdx devices to mount.
> What kernel does it use? Would work native with my SB Audigy?

Linux kernel [4.4.241](https://cdn.kernel.org/pub/linux/kernel/v4.x/linux-4.4.241.tar.xz) (longterm). Here is config. [config.txt](https://github.com/joncampbell123/dosbox-x/files/5504729/config.txt)

Does not have built-in SATA/SCSI/USB etc. support.
Works under - processor family (486).
**Has built-in support:**
-generic ATA/ATAPI disk support
-IDE/ATAPI CDROM
-Intel PIIX/ICH chipsets support (ATA)
-ISA sound devices / Sound Blaster 16
**File systems:** Ext2, ISO 9660, MSDOS fs

Configure the kernel according to your needs and build it.

> Also it has some menu drawing problem that some block of menu background are drawn to the top of screen and doesn't cleared when menu is closed.
> 
This is the mouse cursor pointer.

**Comment by Wengier** [comment]
@grapeli Thanks! I have already posted your reply there.

**Comment by grapeli** [comment]
@Wengier 
Disk Image support.
linux.par needs to be modified (do not blindly copy C/H/S):
```
...
ro
ide-core.chs=0.0:32,16,63
# ide-core.noprobe=0.0
# ide-core.noprobe=0.1
ide-core.noprobe=1.0
ide-core.noprobe=1.1
...
```
More information about [ide-core](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/ide/ide.rst).

dosbox-x (host) config:
```
[ide, primary]
enable = true
[ide, secondary]
enable = true

```
From under dosbox-x
```
imgmount 2 /tmp/hda.img -t hdd -fs none
imgmount d c:\boot.iso -t iso -fs none
linux.bat

```
You should be patient when booting Linux. There is a 30 second pause (hda: interrupt lost).
```
[    0.550794] ide: forcing hda as a disk (32/16/63)
[    0.550794] Probing IDE interface ide0...
[    0.871103] hda: DOSBox IDE disk, ATA DISK drive
[    1.411157] hdb: DOSBox-X Virtual CD-ROM, ATAPI CD/DVD-ROM drive
[    1.431156] ide0 at 0x1f0-0x1f7,0x3f6 on irq 14
[    1.521755] ide: skipping probe for hdc
[    1.531175] ide: skipping probe for hdd
[    1.551157] ide1 at 0x170-0x177,0x376 on irq 15
[    1.551157] ide-gd driver 1.18
[    1.581203] hda: max request size: 128KiB
[    1.581203] hda: 32256 sectors (16 MB) w/2KiB Cache, CHS=32/16/63
[   31.671775] hda: lost interrupt
[   31.771756] ide-cd driver 5.00
[   31.821756] ide-cd: hdb: ATAPI 8X CD-ROM drive, 1536kB Cache
[   31.821756] cdrom: Uniform CD-ROM driver Revision: 3.20

```
From linux.
```
# dmesg
# mount /dev/hda /mnt
# mkdir /tmp/cdrom
# mount /dev/hdb /tmp/cdrom

```
Write-mounted filesystems must absolutely be unmounted before exiting dosbox-x (host).

**Comment by Wengier** [comment]
@grapeli Thanks for the information! I saw he replied the following here to the previous reply:

```
OK, is there some reason why to use modern kernel but stripped by many necessary drivers like SATA and PCI audio subsystem?
If it would be intended for old computers with ATA and ISA sound they don't need DOS emulation and can run programs native...

I could recompile the kernel, I'm not familiar how the dosbox-x image is packed (DOSBOX-X.SQF)
if I could then simply replace the kernel. But currently not much time to play with, I can boot my linux and use dosbox without restrictions...
```
Perhaps things like SATA and PCI audio subsystems as he mentioned can be added and enabled by default for the package so that they can get these to work in DOS when using this approach? And is it possible to configure the disk support more easily (e.g. with an easy-to-use compile script provided if recompile is needed or make it work without the need to recompile if possible)?

**Comment by grapeli** [comment]
@Wengier 
The next person will ask for support for Gravis UltraSound, and another one for Aureal Vortex.
Suggest to install real Linux with Xorg, OpenGL, etc.

Compiling a kernel is very easy.
```
ARCH=x86 make menuconfig
ARCH=x86 make
cp arch/x86/boot/bzImage /tmp/dosbox-x/vmlinuz
```

**Comment by Wengier** [comment]
@grapeli I agree with you that this package is primarily recommended for running in DOSBox-X itself (or qemu as you stated). The official HX-DOS package will be recommended for real DOS systems. And thanks for posting the compiling steps! I have already posted your replies there.

Regarding the mouse cursor pointer issue, do you think it is a bug in the source code or it has something to do with the Linux package? (It does not seem to happen with the HX-DOS package for example)

**Comment by grapeli** [comment]
> Regarding the mouse cursor pointer issue, do you think it is a bug in the source code or it has something to do with the Linux package? (It does not seem to happen with the HX-DOS package for example)

I have no idea what is causing this. I am not a programmer, but a regular user. I noticed that these defects disappear when I choose **Video> Fit to aspect ratio** from the menu.
After you turn off doublescan (**Video> Output> Doublescan**), you can change scalers.

Fbdev is deprecated. The code has been frozen for years. New drivers are not added. It was needed for embedded. KMS replaced him. Its advantage is simplicity (similar to OSS in sound).
`cat /dev/urandom > /dev/fb0`
SDL2 does not have fbcon support directly (via directfb only).

VESA framebuffer has limitations. Screen resolution is rigid.
[Qemu](https://www.kraxel.org/blog/2019/09/display-devices-in-qemu/) supports various frambuffers - vesafb, [cirrusfb](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/fb/cirrusfb.rst), [aty128fb](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/fb/aty128fb.rst). In the latter two, dosbox-x adjusts its resolution on the fly. aty128fb is the fastest.
It's the same with real hardware (framebuffer drivers, better and worse)

Over the years I have used k6-2 350MHz with 3dfx Banshee on Linux. I only used the framebuffer (tdfxfb) to watch movies in mplayer. It had support from MTRR (Memory Type Range Register). Fastest video output (faster than xv). This crappy equipment played DVDs smoothly.

**Comment by Wengier** [comment]
@grapeli You are right that choosing ```Fit to aspect ratio``` from the Video menu will make the defects disappear. In this case, why not make it the default setting for the LOADLIN package so that other users will not encounter this issue by default? It can be set with the config option "aspect=true" in the config file I believe. You can certainly change the default settings for options like doublescan from the config file too.

I have watched the short video showing dosbox-x start in qemu that you posted. It looks like it runs pretty smoothly there indeed.

**Comment by xrayer** [comment]
Hi grapeli, that was me who asked on DOS ain't dead forum and tried your dosbox-x-linux (on a real HW i7-2600k). I can understand that you are just playing with it but why to strip all the working linux drivers and subsystems just to save some MBs of space? I'm not sure what would be a prefered target platform for it. It was proved it can run on old PC with ISA soundcard but it's too slow and useles. For running under QEMU? Why? Where is Qemu there can be run also Dosbox. I just wanted to use it on modern PC from plain DOS without needing to reboot to Windos or Linux so I liked the idea to make a compact dosbox package that can be started from DOS (even better if it can exit back to DOS) and that would emulate SB on my newer SB Audigy PCI (already supported in Linux Alsa).

**Comment by grapeli** [comment]
@xrayer 
I was only curious if it is possible to run DODBox-X in DODBox-X (from linux). Only this. Nothing more. Nothing more.

Why was I also using qemu. Because qemu with kvm is over 100x faster. It is more flexible. Since the handling of disk images is more convenient (no C/H/S need to be specified). Because you can directly mount a directory from the host (`-snapshot -drive file=fat:ro:/tmp/DOS,media=disk,format=raw`). Because the operation of qemu with freedos leaves much to be desired. **DOSBox-X is much better in compatibility with DOS.**

Really don't use this outside of a virtual environment. Dosbox-x for dosbox-x was built for the 486 processor.
`-march=i486 -mtune=generic -mno-mmx -mno-sse -mno-sse2 -mno-sse3 -mno-ssse3 -mno-sse4a -mno-sse4.1 -mno-sse4.2 -mno-sse4 -mno-avx -mno-avx2`
Similarly, the entire buildroot (userland), additionally with the `-Os` optimization. It is slow. Therefore, the fluidsynth is slow and sluggish here.
[Loadlin](https://en.wikipedia.org/wiki/Loadlin) is the bootloader. Install linux (real) do not consider this package as an alternative.

**Comment by xrayer** [comment]
OK, never mind, I just though there would be some use case for a few people who wants to run it from DOS.
I already have Linux and Windows installed in multiboot so I can run Dosbox for Linux/Windows, just having a DOS version (with modern soundcards support) would save a reboot...
I also use Qemu sometimes but it's not 100x faster for dos programs, even using HW virtualization the context switching from VM to hypervisor (VMEXIT), that occurs probably very often, takes thousand cycles that makes things much slower compared to V86 more.

**Comment by grapeli** [comment]
@xrayer 
I used qemu with kvm to test this package because it works over 100 times faster in it.
qemu (kvm) --> linux --> dosbox-x
dosbox-x --> linux --> dosbox-x

This absolutely doesn't mean qemu is 100x faster than dosbox-x in dos, because that's not true. Even if it's faster, it's of little use because too many things don't work as well as they could.

**Comment by Wengier** [comment]
@grapeli DOSBox-X version 0.83.8 is now officially released. Could you make a DOS LOADLIN package for it? Thanks!

**Comment by grapeli** [comment]
I will try to build. New dependencies are coming - freetype (it will take a while).

**Comment by grapeli** [comment]
@Wengier 
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5622556/dosbox-x.linux.zip)
I updated [buildroot](https://buildroot.org/downloads/buildroot-2020.08.2.tar.gz), [kernel](https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.9.11.tar.xz) and [DOSBox-X](https://github.com/joncampbell123/dosbox-x/archive/dosbox-x-v0.83.8.tar.gz).
I haven't tested it very thoroughly. I checked dosbox-x and qemu quickly.
TTF output works.

Dosbox-x is not profiled with PGO (this gives about 8% performance gain). I will update later.

**Comment by Wengier** [comment]
@grapeli Thanks for the updates. I have repacked it here:

https://github.com/joncampbell123/dosbox-x/files/5625195/dosbox-x-dos-0.83.8-loadlin.zip

Compared with the HX-DOS package for 0.83.8, this LOADLIN package supports the new TTF output, which is certainly a plus. Feel free to make further improvements to it. Thanks!

**Comment by grapeli** [comment]
@Wengier 
[Commit](https://github.com/joncampbell123/dosbox-x/commit/b5a582df9d52ae72f398f5bf4e10869178d78223) to turn on automatic date/time synchronization with the host stops linux from starting.
Configuration is required:
```
[dosbox]
synchronize time = false
```

**Comment by Wengier** [comment]
@grapeli The code seems to run fine in Windows, Linux and macOS, so I am not sure what happens with the LOADLIN package in this case. Perhaps you can just disable this option for this package for now, and I will see if there is anything I can do for it.

**Comment by grapeli** [comment]
@Wengier 
Linux booting stops at the earliest stage. `earlyprintk=vga` shows only up to this point.
```
...
...
[0.000000] Initmem setup node 0 [mem 0x0000000000001000-0x00000000057ffffff]
```
Next should be setting the **clock source**.
`[0.000000] clocksource: refined-jiffies:..................`
At first I thought the kernel configuration might have an effect. This one was built without CONFIG_RTC_DRV_CMOS. I've tried different variants already and somehow can't fix it from the linux configuration level.

The most important thing is that synchronization can be turned off from dosbox-x. Let's stay with it.

**Comment by Wengier** [comment]
@grapeli Time synchronization is disabled by default as of the latest code.

Meanwhile, DOSBox-X version 0.83.9 is now released. Could you make a DOS LOADLIN package for it? Thanks!

**Comment by grapeli** [comment]
@Wengier 
I haven't tested too much.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5759306/dosbox-x.linux.zip)

Changed the SquashFS and kernel image compression to zstd.
I added support for several sound cards and some sata/pata drivers (libata) to the kernel.

**Comment by Wengier** [comment]
@grapeli Thanks! I have repackaged it here:

[dosbox-x-dos-0.83.9-loadlin.zip](https://github.com/joncampbell123/dosbox-x/files/5759557/dosbox-x-dos-0.83.9-loadlin.zip)

I think the new package requires the settings "memsize=127" and "cputype=pentium" in order to run.

**Comment by xrayer** [comment]
Hi grapeli, thanks for including the drivers!
Now it detect my disks and SB Audigy. I had to do a trick after boot to make it working - switch to linux console and mount the real partition to /mnt because dosbox-x still cannot mount it - complains that cannot mount physFS, then I need to mount it in dosbox window e.g. "mount d /mnt". Also I had to run alsamixer to boost master volume that was 0 by default.
Could I make some changes permanent on booted system or do I need to mount DOSBOX-X.SQF in other linux and edit settings?
Then I could play games on my dos partition BUT when any game or my VESATEST tool try to switch in VESA LFB mode the whole thing hung - I mean also the host linux and I had to press reset button. Games like NFS1 that probably don't use LFB worked in SVGA mode.

**Comment by grapeli** [comment]
@xrayer 
You have to make these changes yourself. My test environment is limited to dosbox-x and qemu.
It is not difficult. Use some relatively recent Linux distribution. Install it anywhere, in a virtual machine, etc.
Install squashfs-tools with zstd support.
Unpack the rootfs:
```
unsquashfs dosbox.sqf
cd squashfs-root
```
Make the necessary changes. For example, the `alsactl store` or `tinymix` command is used to change the mixer settings.
Almost all configuration files are in `/etc` (/etc/inittab, /etc/init.sh, /etc/fstab). This Linux is simplified to the maximum.
Pack again.
`mksquashfs squashfs-root /tmp/dosbox.sqf -comp zstd -b 256K -no-recovery -all-root -no-xattrs -not-reproducible -no-exports -all-time $(date +%s) -Xcompression-level 22`

**Comment by xrayer** [comment]
OK, I already tried under my not so updated Debian via mount command and it wrote that FS has unknown compression. I'll try to update some package and try your way.

I also found that in higher VESA mode 1280x1024 or 1600x1200 the Quake start in VESA mode but when switching to other modes the full freeze happens again. 
BTW where is defines the resolution of Dosbox GUI? When I set FB to 1600x1200 the dosbox seems to use 800x600 (small in the middle of screen) but when start e.g. quake in 1024x768 dosbox window will resizes up.

Another issuse had with new win32 dosbox-x 0.83.9 that doesn't run under WinXP - PE header need to be fixed: OS version 6.0 -> 5.1 then it runs but when packed by UPX it again complains "invalid win32 app" but I could compress with pklite32w with worse compression ratio (7,5MB intead of 5,1MB)...

**Comment by xrayer** [comment]
I got:
root@MR:/tmp/X# unsquashfs DOSBOX-X.SQF 
Filesystem uses unknown compression, this is unsupported by this version
Decompressors available:
	gzip
	lzo
	xz
Can I easily add your used compression method via APT? (Debian 8.11 jessie)

**Comment by grapeli** [comment]
> It is not difficult. Use some relatively recent Linux distribution.

@xrayer 
Debian jessie does not fall into this category. It's an old, unsupported system.
I prepared you a version with gzip compression.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5760695/dosbox-x.linux.zip)

**DOS version of DOSBox-X does not run in DOSBox-X.** 
It doesn't do any more, and I don't generally provide any other support.

**Comment by xrayer** [comment]
Thanks, in mean time I was able to update zstd and squashfs packages so I'm able to decompress/compress.

I already fixed mixer setting, changed some in dosbox.conf but I have problem with automounting. When I added into /etc/fstab lines like this
/dev/sdb9     /mnt/i      msdos  rw
It was not automounted on start and when I manualy type 
mount /mnt/i
I got error: can't find /mnt/i in /etc/fstab
I could add mount commands in init.sh but why fstab way is not working?

I also found that reboot host os command in menu doesn't work, neither press CTRL+ALT+DEL in console do nothing, I have to type reboot command.

**Comment by xrayer** [comment]
(the /mnt/i and other subdirs are created and exists, when I type mount /dev/sdb9 /mnt/i it gets mounted normally but mount -a ignore it)

**Comment by grapeli** [comment]
@xrayer 
In the /etc/inittab file you have an commented out (disabled) function. Uncomment.
`::ctrlaltdel:/sbin/reboot`

As for the non-functioning `fstab` filesystem mount. I assure you that the correct entry will work. Yours is incomplete.

**Comment by xrayer** [comment]
Oh I missed 0, 0 at end for backup and fsck...
Now automount works, also I enabled the CTRL+ALT+DEL reboot.
Now it's well setup for instant DOS usage :)
Except I worry about freezing during VESA modes change but I know it may be harder to debug...

**Comment by grapeli** [comment]
@xrayer 
The Linux VESA framebuffer has limitations. It has a fixed screen resolution. It can be changed at startup kernel parameter (linux.par) vga=791
```
vga=785  #640x480 16bit
vga=788  #800x600 16bit
vga=789  #800x600 24bit
vga=791  #1024x768 16bit
vga=794  #1280×1024 16bit
vga=795  #1280×1024 24bit

```
It should also be possible to change with the `fbset` command after booting (unfortunately it does not work). There should be the appropriate database in /etc/fb.modes.
https://linux.die.net/man/5/fb.modes
It can be generated under linux with the command `gtf`.
`gtf 800 600 75 -f`
or
`gtf 800 600 75 -f > fb.modes`

Coming back to the vesa framebuffer.
Dosbox-x running under it cannot smoothly adjust the required resolution. It only uses the one that was set earlier. Not less than 800x600. It is not perfect.

Say what graphics card you are using on this hardware. Maybe there is a better framebuffer than VESA for it.

**Comment by xrayer** [comment]
I though that Linux FB resolution doesn't need to be changed and it already doesnt. I set e.g .795 and get Dosbox centered view, 800x600, when start quake with 1024x768 as default then Dosbox window will resize up but still keeps native res. 1280×1024 of the FB so I guess that Dosbox handle the transformation from guest to host some way and the modes that dosbox offers to quest are not linked with real VESA modes of host card. When I switch quake to lower or higher res. I would expect the dosbox window will resize but still keeps in same FB mode but it freeze instead. I don't know the internal graphics processing but it seems that some nasty things are going there when it can corrupt the linux behind - only the dosbox should crash if some problem...
I currently use nVidia GTX670 so alternative may be nouveau...

**Comment by grapeli** [comment]
This kernel config is generic, tailored to the very small requirements of running DOSBox-X in DOSBox-X (qemu).
I think it completely does not use the potential of your equipment. I thought you might be running it on hardware older than 2002.
This package doesn't take advantage of your hardware, mainly your graphics card (cpu probably too).
Not using OpenGL!!!

If it suits you.
Run linux "live cd" (without GUI) and send the result of the lsmod command.

**Comment by grapeli** [comment]
https://en.wikipedia.org/wiki/GeForce_600_series
Your graphics card is compatible with OpenGL 4.6 and Vulkan 1.1. My of 2.1. Vulkan does not support.

This package will not use it in any way. You can possibly use it to create a new one - SDL2 with the KMS-drm backend. He uses OpenGL (without installing Wayland or Xorg).

Install a full-fledged operating system there (Windows, linux or other) and run DOSBox-X on it.

**Comment by xrayer** [comment]
Of course I know that VESA doesn't offer any acceleration but for most of older low-res games it's fine. I already have multiboot with linux and windows so I can run dosbox there but this version saves some times to reboot, only then reboot back to dos. BTW the performance difference is not so extremely different - a timerefres in Quake @1024x768 unde this dosbox gave ~18FPS and under Windows I got ~30FPS.

**Comment by grapeli** [comment]
I removed VESA fb and added nouveaufb.
Try with this kernel. Maybe it will function better.
[linux-4.19.164.zip](https://github.com/joncampbell123/dosbox-x/files/5776623/linux-4.19.164.zip)

**Comment by xrayer** [comment]
Thanks, I tried it but got just kenel panic soon at boot
http://rayer.g6.cz/1tmp/nvfb.jpg

**Comment by grapeli** [comment]
@xrayer 
Provide information about your equipment. Are you using this motherboard?
https://www.gigabyte.com/Motherboard/GA-P67-DS3-B3-rev-10#ov
She is also EFI compliant. What CPU are you using.

Run some new (fresh) Linux live distribution on that hardware and give at least `lsmod` output.

**Comment by grapeli** [comment]
In the meantime, I prepared the x86_64 version. I suspect you are definitely using a compatible processor.
Kernel, Buildroot, and DOSBox-X are 64-bit. It works about 7-9% faster under Qemu with some benchmarks (pcbench, dhry1nd, doom, quake).
This is the VESA fb version. We'll work on the nvidiafb (nouevaufb) version later.
[dosbox-x.linux64.zip](https://github.com/joncampbell123/dosbox-x/files/5787207/dosbox-x.linux64.zip)

**Comment by xrayer** [comment]
Yes the MB is Gigabyte GA-P67-DS3-B3 but the BIOS is one of latest classic Award BIOS - No UEFI.
CPU: intel Core i7-2600K SandyBridge, 4GB RAM. In my Debian install I use debianized nVidia proprietary drivers package fo X but console with VESA, not sure now, test it later...

**Comment by grapeli** [comment]
One more 64-bit kernel. Without VESA fb but with i915drmfb and nouveaufb.
[linux-x86_64-4.19.165.zip](https://github.com/joncampbell123/dosbox-x/files/5789177/linux-x86_64-4.19.165.zip)

The default fb is likely to be an integrated Intel GPU. View in dmesg. You can also check like this:
`cat /sys/class/graphics/fb0/name`
and
`cat /sys/class/graphics/fb1/name`
Dosbox-x on the second fb can run like this.
`SDL_FBDEV=/dev/fb1 dosbox-x`

This dosbox-x app is really pretty fast. I ran it at home on a much weaker i5-450M processor (slower than yours at least ~46%, per core).

My fb is `i915drmfb`.
dhry1nd - 114-120
pcpbench - 76 (640x400)
quake (with sound) 
1024x768 - 19.4
800x600  - 29.8

It's a bit slower under Qemu.

**Comment by xrayer** [comment]
Ah, I just tested previous x64 version and see there's another one, you do so fast :)
It rocks, got twice FPS in Quake :) VESA mode switching almost worked except when I enabled 1600x1200 in dosbox.conf and then tried to set it freeze. Also one time it freeze at exit from doom before it should switch back to text mode when I played with scalers before. 2nd time when didn't touch scalers doom exited normally.
lsmod here http://rayer.g6.cz/1tmp/lsmod.txt

**Comment by grapeli** [comment]
Doublescan must be disabled when using scalers.

There is no sysrq running on this kernel. I will add, because it is necessary, when the application hangs on the framebuffer then you can return to the console, `alt-SysRq-v` (otherwise only via ssh, etc.).
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/admin-guide/sysrq.rst#n138
https://en.wikipedia.org/wiki/Magic_SysRq_key

**Comment by xrayer** [comment]
Hi, I tried the version linux-x86_64-4.19.165.zip
cat /proc/fb shows nouveaufb
I don't have any problems with switching lower modes. 1600x1200 still freeze, OK it's some bug as you described, at least it should crash the dosbox on some memory protection error instead freeze whole kernel. But it's not big limitation for me. Speed seems to be same as previous.

**Comment by Wengier** [comment]
@grapeli DOSBox-X version 0.83.10 was just released. Could you make a DOS LOADLIN package for it? Thanks again.

**Comment by grapeli** [comment]
@Wengier 
I added libpng. It is possible to take screenshots.
It can be displayed after exiting dosbox-x with `fbv /tmp/capture/*.png`.
`cpu=pentium` is required.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/5904000/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Great! It certainly works and is already available for download from the website.

**Comment by Wengier** [comment]
@grapeli DOSBox-X version 0.83.11 is now released. Could you please make a DOS LOADLIN package for it as before? Thanks!

**Comment by grapeli** [comment]
@Wengier 
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/6064059/dosbox-x.linux.zip)
Generally no significant changes.
Record video to AVI (ZMBV) does not work. It looks like regressions. It's not just this special version. Attempting to capture video to avi (zmbv) under dosbox-x-0.83.11 in archlinux ends with segfault.

**Comment by grapeli** [comment]
@Wengier 
Sorry to mess it up. False alarm.
I will fix and update the dosbox-x version launched by loadlin (record to avi doesn't work here).

**Comment by Wengier** [comment]
@grapeli Thanks for the update. The package has been updated accordingly.

**Comment by grapeli** [comment]
@Wengier 
The newest version, 0.83.12.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/6246161/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Thanks! The download link now appears on the homepage.

**Comment by xrayer** [comment]
Would be there also 64-bit build that I could upgrade that experimental version with Alsa&nouvaufb made before?

**Comment by grapeli** [comment]
@xrayer 
I accidentally deleted my previous kernel configuration.
Test this build.
[dosbox-x-x86_64.zip](https://github.com/joncampbell123/dosbox-x/files/6251268/dosbox-x-x86_64.zip)

For me it works pretty fast under `i915drmfb`. It works three times slower under `vesafb`.
Only using SDL2 and kmsdrm + opengl can solve the scaling problem.

**Comment by xrayer** [comment]
I think the kernel config file can be extracted from kernel imge, if it was compiled with it, so I could extract it. I'll test also nev ver.

**Comment by Wengier** [comment]
@grapeli DOSBox-X version 0.83.13 is now released. Can you make a new DOS LOADLIN package for it? Thanks a lot!

**Comment by grapeli** [comment]
@Wengier
The newest version.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/6409149/dosbox-x.linux.zip)

There are two trivial bugs when compiling dosbox-x with the musl-libc based toolchain ([buildroot](https://buildroot.org/)). Related to header files. I do not report them because apparently nobody but me uses musl libc to compile dosbox-x (I can live with them).
In the [sdl_mapper.cpp](https://github.com/joncampbell123/dosbox-x/blob/master/src/gui/sdl_mapper.cpp) file:
missing `#include <limits.h>` which is required explicitly on musl-libc.
```
sdl_mapper.cpp: In function ‘void MAPPER_SaveBinds()’:
sdl_mapper.cpp:4143:15: error: ‘PATH_MAX’ was not declared in this scope
 4143 |     char path[PATH_MAX];
      |               ^~~~~~~~
sdl_mapper.cpp:4144:43: error: ‘path’ was not declared in this scope
 4144 |     if (realpath(mapper.filename.c_str(), path) != NULL) LOG_MSG("Saved mapper file: %s", path);
      |                                           ^~~~
sdl_mapper.cpp: In function ‘void ReloadMapper(Section_prop*, bool)’:
sdl_mapper.cpp:4878:7: warning: this ‘else’ clause does not guard... [-Wmisleading-indentation]
 4878 |     } else
      |       ^~~~
sdl_mapper.cpp:4880:2: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘else’
 4880 |  if (init) {
      |  ^~

```

And in the [porttalk.cpp](https://github.com/joncampbell123/dosbox-x/blob/master/src/libs/porttalk/porttalk.cpp) file, there must be `#include <sys/io.h>` instead of `#include <sys/perm.h>`.

**Comment by Wengier** [comment]
@grapeli Thanks! The updated DOS LOADLIN package already appears in the website. I never used musl-libc based toolchain myself in any OS, including Fedora Linux. But I think ```#include <limits.h>``` can certainly be added as you mentioned for greater compatibility.

As for the porttalk.cpp file, I believe you are saying that #include <sys/perm.h> need to be replaced by #include <sys/io.h> for the musl-libc based toolchain, although I am not quite sure if this will break anything else. Does include both headers work in your case?

**Comment by grapeli** [comment]
> As for the porttalk.cpp file, I believe you are saying that #include <sys/perm.h> need to be replaced by #include <sys/io.h> for the musl-libc based toolchain, although I am not quite sure if this will break anything else. Does include both headers work in your case?

@Wengier 
Unfortunately not.
```
porttalk.cpp:243:10: fatal error: sys/perm.h: No such file or directory
  243 | #include <sys/perm.h>
      |          ^~~~~~~~~~~~
compilation terminated.
```
https://git.musl-libc.org/cgit/musl/tree/include/sys/io.h

**Comment by Wengier** [comment]
@grapeli I am not personally sure how to handle this case then. Unless ```#include <sys/io.h>``` works for other cases as well, perhaps there is some way to check if the musl-libc based toolchain is used, and include the appropriate header accordingly?

**Comment by grapeli** [comment]
@Wengier 
Also, I don't know how this could be solved. I did not googgle anything similar.

Linux distributions (Alpine, Void, etc.) based on musl libc are full of special patches to adapt the software to the needs of this other C base library. There is nothing wrong with applying trivial patches.

**Comment by grapeli** [comment]
@Wengier 
I'm adding dosbox-SVN-r4450, which can run in DOSBox-X without any problems.
VGA is over 3x faster (than dosbox-x). CPU emulation similar. This is best seen when using the TTF output (it is definitely the fastest).
The startup time is shorter, among other things, due to the much smaller size (unpacking is faster). 
`./dosbox-x dosbox-svn.zip -set memsize=64 -set cputype=pentium -set core=dynamic_x86 -set cycles=max -c "c:" -c "linux.bat" 2>/dev/null`
[dosbox-svn.zip](https://github.com/joncampbell123/dosbox-x/files/6410110/dosbox-svn.zip)

**Comment by Wengier** [comment]
@grapeli Can you give some information on what may be the cause(s) that impact performance other than the unpacking size? DOSBox-X tries to implement emulation as seen in real DOS systems, but in many cases it may just be some settings (like “hard drive data rate limit”) or inefficient code (like GetLongName function) which cause speed concerns (rather than the emulation itself). In such cases the issues can be resolved pretty easily without any other impacts to the emulation itself, and such information can be added to documentation as well for people who need them.

**Comment by grapeli** [comment]
@Wengier 
 The difference is considerable in action.
Running under qemu (kvm) the difference is small.
Although for me under Linux, of course, this is not the case in the real hardware. Yes, in general I can say that SVN is a bit faster. Very similar in some (maybe the most common),  but I can give you some tests where it is slower.
```
export SDL_VIDEODRIVER=dummy
strace dosbox -c exit 2>&1 | wc -l
735
export SDL_VIDEODRIVER=dummy
strace dosbox-x -c exit 2>&1 | wc -l
1182
```
The start time is the difference in the number of system calls and file access. Dosbox-x is linked to other libraries - libfluidsynth.so, libfreetype.so, libglib-2.0.so, libpcre.so, libffi.so. Each of them must be read (unpacked).  The difference in size alone is over 7MB. All rootfs are compressed (zstd). The difference in unpacking with such modest computing power is visible.

Run dosbox-x in dosbox-x. Set ttf output (in guest). Run `dhry1nd`. Then `pcpbench`. Why does `pcpbench` start fast (ttf) and sluggish on surface video output?
I guess the difference lies in the slower graphics emulation. With such low computing power, this difference becomes more acute.

**Comment by Wengier** [comment]
@grapeli Thanks for the detailed information. According to this TTF is the fastest output among all. Not sure exactly why though.

Meanwhile, I'm glad to mention that I found that with DOSBox-X version 0.83.13 or latest code, HX-DOS version of DOSBox-X does appear to launch inside DOSBox-X itself with the default setting and `-memsize=127`:

![image](https://user-images.githubusercontent.com/8216923/119406617-10adf000-bcb1-11eb-86dc-4865d3931657.png)

But it does not appear to work realizably. For example, the mouse does not work well at all, and typing EXIT will cause it to crash. But it finally runs after all.

**Comment by grapeli** [comment]
@Wengier 
The easiest way to capture the difference between `ttf` output and `surface` output is with `SDL_VIDEODRIVER=dummy`.  

For example.
https://fractint.org/
```
time SDL_AUDIODRIVER=dummy SDL_VIDEODRIVER=dummy ./dosbox-x -fastlaunch -set "dos hard drive data rate limit=0" -set showmenu=false -set nosound=true -set mididevice=none -set sbtype=none -set pcspeaker=false -set output=ttf -set doublescan=false -set core=dynamic_x86 -set cycles=max -c "mount c ." -c "c:" -c "fractint @FRACT19.PAR/NEWTON-REAL video=af4 textsafe=no virtual=no batch=yes overwrite=yes" -c exit 2>/dev/null
time SDL_AUDIODRIVER=dummy SDL_VIDEODRIVER=dummy ./dosbox-x -fastlaunch -set "dos hard drive data rate limit=0" -set showmenu=false -set nosound=true -set mididevice=none -set sbtype=none -set pcspeaker=false -set output=surface -set doublescan=false -set core=dynamic_x86 -set cycles=max -c "mount c ." -c "c:" -c "fractint @FRACT19.PAR/NEWTON-REAL video=af4 textsafe=no virtual=no batch=yes overwrite=yes" -c exit 2>/dev/null
```

Linux perf shows where the difference comes from. With the involvement of two additional functions for the `surface` output.
```
2.06%  dosbox-x                    [.] VGA_TEXT_Xlat32_Draw_Line
0.90%  dosbox-x                    [.] RENDER_StartLineHandler
```
If the test run takes longer, the difference will be more noticeable, even by 7%.

**Comment by Wengier** [comment]
@grapeli Thanks for the information. Just let you know that DOSBox-X does support the -silent command-line option, which implies `SDL_VIDEODRIVER=dummy`, `-fastlaunch` and `-c exit` so that the command lines can be made shorter I think:

`time SDL_AUDIODRIVER=dummy ./dosbox-x -set "dos hard drive data rate limit=0" -set showmenu=false -set nosound=true -set mididevice=none -set sbtype=none -set pcspeaker=false -set output=ttf -set doublescan=false -set core=dynamic_x86 -set cycles=max -c "mount c ." -c "c:" -c "fractint @FRACT19.PAR/NEWTON-REAL video=af4 textsafe=no virtual=no batch=yes overwrite=yes" 2>/dev/null`

Perhaps it should imply `SDL_AUDIODRIVER=dummy` as mentioned too.

**Comment by grapeli** [comment]
@Wengier 
Thanks for this info. I will use the abbreviated form.

One more remark. Maximizing performance with silent mode (`SDL_VIDEODRIVER=dummy`) and `ttf` output always takes place.
In the case of a visible window, I'm not sure. It can be different. Maybe it depends on where this window is displayed - windows, macosx, linux (X11, wayland, Xwayland). Perhaps it is worth turning off the blinking cursor, `ttf.blinkc=false` (this also generates minimal events).

**Comment by Wengier** [comment]
@grapeli It is strange that when I tried the timing with -silent in MinGW builds, the TTF output seems to take more time to complete than the default one. So I guess it may indeed depend on the platform too, i.e. faster in some but maybe slower in others.

**Comment by grapeli** [comment]
@Wengier 
Batch - compilation (watcom) and various types of compression and decompression.
DOSBox-X linux, SDL2. All runs with nosound=true.

| x  | i3 (xorg) |
| ------------- | ------------- |
| output=surface  | 8m47.512s |
| output=opengl  | 8m44.353s |
| output=ttf, ttf.blinkc=false | 8m6.522s |
| output=ttf, ttf.ptsize=9, ttf.blinkc=false | 8m5.598s |
| **silent, output=ttf** | **8m0.583s** |
| silent, output=surface | 8m31.784s |

You can clearly see that the ttf output is the fastest. 9% is a lot.

**Comment by Wengier** [comment]
@grapeli Thanks for the comparison table. I think I probably see where the main difference comes from: if I just launch DOSBox-X with -silent and output=ttf without doing non-trivial things, then output=surface or output=default is perhaps faster than output=ttf. But if there are a lot of outputs like operations mentioned above, then -silent and output=ttf is probably much faster than output=surface or output=default. So it is certainly recommended to use output=ttf in combination with -silent if there are a lot of operations like above, but output=default may be recommended for -silent if the user just wants to run some relatively trivial stuff silently. I will mention this in the Wiki too. Thanks!

**Comment by Wengier** [comment]
@grapeli DOSBox-X 0.83.14 is being released. Can you make a DOS LOADLIN package based on the latest code? Thanks!

**Comment by grapeli** [comment]
@Wengier 
I added the Terminus font as the default. After exiting dosbox-x to activate the virtual console, press Enter or switch to the other Alt-F2. This time compiled with LTO. It's not faster, but it's smaller. Surprisingly, it compresses less, the squashfs image is larger.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/6575047/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Thanks for the update. It already appears in the website. I did not use the one you said has "drawback" yet because the drawback as you described may be an important one (for the menu). If this can be fixed then it would be great, or do you mean it is simply impossible to fix?

**Comment by grapeli** [comment]
@Wengier 
This is a limitation of the `vesafb` framebuffer. It cannot change the resolution. You can only declare the required one when booting. For dosbox-x in dosbox-x, the `s3fb` [framebuffer](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/tree/drivers/video/fbdev/s3fb.c?h=linux-4.19.y) is the best choice. There are no such restrictions. Next time I will try to prepare such a version.

**Comment by BridgeHeadland** [comment]
There's one thing I'm wondering, I realize it may seem strange that it's coming from me: Why run DOSBox in DOSBox really?

**Comment by grapeli** [comment]
@BridgeHeadland 
Type of test - compatibility, compliance. Nothing more.

I'll reverse the question. Why are you running Doom Legacy in DOSBox-X?

**Comment by BridgeHeadland** [comment]
> @BridgeHeadland
> Type of test - compatibility, compliance. Nothing more.
> 
> I'll reverse the question. Why are you running Doom Legacy in DOSBox-X?

I don't running Doom Legacy in DOSBox-X. I just run iwads in DOSBox sometime by type for example "doom2 -file NERVE.WAD" because I like to run any wads with the low 320x200 resolution, expect TNT: Evilution and The Plutonia Experiment, where I prefer to run them with the 640x480 resolution in Doom95 (Final Doom).
Over to something else: Is there any difference between running DOSBox-X for DOS in DOSBox-X for Windows and running DOSBox-X for Windows without running DOSBox-X for DOS there? I'm not particularly familiar with that.

**Comment by Wengier** [comment]
@BridgeHeadland Not all features in DOSBox-X for Windows can be supported by DOSBox-X for DOS, because of some apparent limitations there. But the core functions of DOSBox-X should work just fine in the DOS version.

**Comment by BridgeHeadland** [comment]
> @BridgeHeadland Not all features in DOSBox-X for Windows can be supported by DOSBox-X for DOS, because of some apparent limitations there. But the core functions of DOSBox-X should work just fine in the DOS version.

When I try to run DOSBox-X for DOS in DOSBox-X for example for Windows (dosbox-x.exe), I get the message: "video mode 800x600x16 with LFB not supported", what does it mean?

**Comment by grapeli** [comment]
I do not know. There are two commands in the terminal under Linux.
```
curl -JOL https://github.com/joncampbell123/dosbox-x/releases/download/dosbox-x-v0.83.14/dosbox-x-mingw-hx-dos-20210601013916.zip
dosbox-x -set memsize=127 dosbox-x-mingw-hx-dos-20210601013916.zip -c "c:" -c "dosbox-x.exe"
```
I do not see such a message. DOSBox-X runs in DOSBox-X.

**Comment by BridgeHeadland** [comment]
> I do not know. There are two commands in the terminal under Linux.
> 
> ```
> curl -JOL https://github.com/joncampbell123/dosbox-x/releases/download/dosbox-x-v0.83.14/dosbox-x-mingw-hx-dos-20210601013916.zip
> dosbox-x -set memsize=127 dosbox-x-mingw-hx-dos-20210601013916.zip -c "c:" -c "dosbox-x.exe"
> ```
> 
> I do not see such a message. DOSBox-X runs in DOSBox-X.

![DOSINDOS](https://user-images.githubusercontent.com/85258509/121932127-c9ad9a80-cd44-11eb-9cfa-3b15f10bae4f.PNG)

**Comment by joncampbell123** [comment]
@BridgeHeadland Are you running DOSBox-X for DOS in 32-bit Windows? That involves NTVDM which does not emulate the VESA BIOS extensions as far as I know. It did implement them in Windows XP, but as far as I know that was removed in Windows Vista because it doesn't fit into the new display driver model and who runs graphical DOS stuff anymore anyway, right?

The only other reason it would say that is if you are running it on real DOS hardware with a video card that is so old, it pre-dates VESA BIOS extensions or a linear framebuffer. Almost any video card since mid 1990s VLB and PCI should have a VESA BIOS and a linear framebuffer.

**Comment by grapeli** [comment]
@Wengier 
The newest version.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/6750383/dosbox-x.linux.zip)

Small compilation error.
```
int_dosv.cpp:170:15: error: ‘PATH_MAX’ was not declared in this scope
  170 |    char cname[PATH_MAX + 1];
      |               ^~~~~~~~
int_dosv.cpp:171:12: error: ‘cname’ was not declared in this scope; did you mean ‘fname’?
  171 |    sprintf(cname, ".%s", start);
      |            ^~~~~
      |            fname
```
The `limits.h` header file is missing from the `dosv.cpp` file.

**Comment by Wengier** [comment]
@grapeli Thanks! The said header file was added and the updated DOS LOADLIN package now appears on the website.

**Comment by grapeli** [comment]
@Wengier 
Build 0.83.16.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/6914149/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Thanks. It has been repackaged as usual and appears in the website.

**Comment by Wengier** [comment]
@grapeli Is there a DOS loadlin package for the 0.83.19 version? Thanks.

**Comment by grapeli** [comment]
@Wengier I will try to prepare. In about 12 hours.

**Comment by grapeli** [comment]
Build 0.83.19.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/7459222/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
Thanks @grapeli. It loads DOSBox-X fine.

![image](https://user-images.githubusercontent.com/8216923/139818979-67f088ba-ba24-4105-a211-a5754adc48fa.png)

**Comment by Wengier** [comment]
@grapeli DOSBox-X 0.83.20 is now released. Is there a DOS LOADLIN package for it? Thanks.

**Comment by grapeli** [comment]
Build 0.83.20.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/7632660/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Thanks. It is now up.

**Comment by Wengier** [comment]
@grapeli DOSBox-X 0.83.21 is now released. Do you have a DOS LOADLIN package for it? Thanks.

**Comment by grapeli** [comment]
@Wengier 
> By the way, can you add language files to the file system so that they can be loaded?

Next time.

Build 0.83.21.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/7797573/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli It is up now.

**Comment by grapeli** [comment]
@Wengier 
> @grapeli It is up now.

Very good news for the beginning of 2022.
I didn't add language files because the whole thing was to be as small as possible. Buildroot is built without NLS (Native Language Support) without locale support, moreover its essential part of busybox without unicode support.

One more observation. More and more system dialogs are added to dosbox-x. When running dosbox-x on `fbcon` or `kmsdrm` (**it is possible that someone will run it on RPi**, for example), this pop-up is blocked (it's good to know [magic SysRq](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/tree/Documentation/admin-guide/sysrq.rst?h=linux-4.19.y) shortcuts). Remember when using `kmsdrm`, `fbcon` you cannot use anything else at the same time.

For example, let's take this [commit](https://github.com/joncampbell123/dosbox-x/commit/e8aea288760b523e5ec1a6d063b76d8d73c8a84f).
Run this dosbox-x in dosbox-x or wherever you want.
Make a screenshot. OKAY. Works fine (because dosbox-x is invoked from init).

Quit dosbox-x. Hit enter. Run now directly from the shell.
Make a screenshot. Suddenly everything freezes. Tiny file dialogs appears underneath, waiting for "Enter" to be pressed. You are blocked at this point.
Try again. Now preventing the unwanted "window" from appearing.
`( dosbox-x & ) >/dev/null`

**Comment by Wengier** [comment]
@grapeli I tried to see the CJK characters as displayed by the TTF output in the LOADLIN build, which was why I asked to add the language files (otherwise they are not accessible from the file system). The TTF output does work in the LOADLIN build as far as I know.

**Comment by grapeli** [comment]
@Wengier 
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/7798880/dosbox-x.linux.zip)

I added these language files. They are located in `/usr/share/dosbox-x/translations`.

Of course, you can attach a disk image (I am thinking of running dosbox-x in dosbox-x).
The kernel has added support for such filesystems, among others - `ext2`, `vfat`, `msdos` (`9p` from Qemu).
For `hd_520`. In the `linux.par` file, change the lines to:
```
#ide-core.chs=0.0:32,16,63
ide-core.chs=0.0:489,16,63
```
**Please mount this disk image before starting loadlin.**
`imgmount 2 hda.img -size 512,63,16,489 -t hdd -fs none`
Be patient. There will be a 30-second pause that cannot be skipped - "hda: lost interrupt" (unless you use Linux 2.2.x from 1999).
Quit dosbox-x. Mount filesystem from shell.
`mount /dev/hda1 /mnt`
Run dosbox-x.
`dosbox-x -set output=ttf -lang /mnt/src/dosbox-x/contrib/translations/es/es_ES.lng -c "mount d /mnt/dos"`

**Comment by Wengier** [comment]
@grapeli Can you make a DOS LOADLIN package for the latest release (0.83.23)? Thanks!

**Comment by grapeli** [comment]
@Wengier 
I will try to prepare (in a few hours).

**Comment by grapeli** [comment]
@Wengier 
Build 0.83.23. I only checked it briefly in action. I haven't tested it.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/8189441/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli Thanks. Added the link accordingly.

**Comment by grapeli** [comment]
Boot up properly with `cputype=pentium`,` pentium_ii`, `ppro_slow` (host settings).
Does not work properly with `cputype=pentium_iii`.

https://user-images.githubusercontent.com/452325/156884666-9fee1d84-0df2-49e6-a676-fd8d17e9a635.mp4

**Comment by Wengier** [comment]
@grapeli DOSBox-X 0.83.24 is being released. Can you make a DOS LOADLIN package for it? Thanks!

**Comment by grapeli** [comment]
@Wengier
Build 0.83.24.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/8395634/dosbox-x.linux.zip)

**Comment by Wengier** [comment]
@grapeli DOSBox-X 0.83.25 has been just released. Can you make a LOADLIN package so that it can be run in DOSBox-X itself? Thanks!

**Comment by grapeli** [comment]
@Wengier 
Build 0.83.25.
[dosbox-x.linux.zip](https://github.com/joncampbell123/dosbox-x/files/8599508/dosbox-x.linux.zip)

It works. Thanks to the new option `show recorded filename=false`, even better, it does not lock up when taking a screenshot, for example.

**Comment by Wengier** [comment]
@grapeli Thanks for it. I can confirm it can run as expected.

**Comment by BridgeHeadland** [comment]
When I start dosbox-x-mingw-hx-dos-20221003161227 (HX-DOS) with the executable file dosbox-x.exe, in dosbox-x-vsbuild-win64-20221003161214 (DOSBox-X for Windows), I get the message " DPMILD32: C:\DOSBOX-X.EXE: out of memory". What can come of it? I have 72.2 GB of space left on the hard drive of my PC with Windows 10 (which belongs to the Windows 1x family, one might say 😉) as an operating system. Could it matter? Is 72.2 GB really too little memory for HX-DOS, or is there perhaps another reason?
I'm attaching the configuration files from HX-DOS, with my custom settings (which I consider perfect and logical settings), maybe someone here can see what the problem is, and maybe someone here has some questions regarding my configuration settings.
[HXDOS.zip](https://github.com/joncampbell123/dosbox-x/files/9748213/HXDOS.zip)

**Comment by BridgeHeadland** [comment]
@joncampbell123 At least I don't have NTVDM on my PC, so I wonder if the configuration files, which come with HX-DOS, really have any effect on HX-DOS, since they come with it. I would like to see if I can run HX-DOS with Tandy as machine value, in DOSBox-X with VESA 2.0 (NOLFB) as machine value.
At least I can find out here, without having to test it myself, not until there is a replacement or something for NTVDM, if that happens.
Regarding HXGUIHLP.DLL, why is the default setting at 800x600x16? Why isn't it at -1x-1x-1 - which (at least in my case) is at 4294967295x4294967295x4294967295? I have tried 0x0x0 to, but then no picture comes up.

**Comment by grapeli** [comment]
[ScummVM 2.6.1](https://drive.google.com/file/d/1GLC1rlaeIvDvwYOatUDPlm0gV16ADh71/view?usp=sharing) for DOSBox-X.

I run it like this.
`dosbox-x -set "dos floppy drive data rate limit=0" -set "dos hard drive data rate limit=0" -set cputype=pentium -set memsize=128 -set cycles="max 105%" -set core=dynamic_x86 -set output=opengl -c "imgmount 0 ./boot/grub.ima -t floppy -nofs" -c "mount c ." -c "boot a:"`

For what purpose? Neither. Just for fun. Test version. Certain things certainly don't work.
With fluidsynth and mt32. I do not suggest activating them, they need much more cpu power than emulation can provide.

Keyboard shortcuts:
Ctrl+m  displays the global main menu.

There is still a version under win9x, I didn't try to run it under dosbox-x.
